src/webnn/native/dml/DMLUtils.cpp

```diff
@@ -52,42 +61,396 @@ namespace webnn::native::dml {
+                            outputArrayBufferView.byteLength != 0);
+            } else {
+                size_t byteLength = reinterpret_cast<const DML_BUFFER_TENSOR_DESC*>(
+                                        outputNodes[i].outputTensorDesc.Desc)
+                                        ->TotalTensorSizeInBytes;
+                // It is an unuseful output of dml graph. We need not read back and copy buffer
+                // to it, just reserve it as a placeholder.
+                outputArrayBufferView = {nullptr, byteLength, 0};
+                outputArrayBufferViews.push_back(outputArrayBufferView);
+            }
+        }
+
+        std::vector<DML_BINDING_DESC> outputBindingDesc(outputNodes.size());
+        std::vector<DML_BUFFER_BINDING> outputBufferBinding(outputNodes.size());
+
+        uint64_t outputOffset = 0;
+        for (size_t i = 0; i < outputNodes.size(); ++i) {
+            auto output = outputArrayBufferViews[i];
+            outputOffset =
+                RoundUpToMultiple(outputOffset, (uint64_t)DML_MINIMUM_BUFFER_TENSOR_ALIGNMENT);
+            outputBufferBinding[i].Buffer = mOutputResource.Get();
+            outputBufferBinding[i].Offset = outputOffset;
+            outputBufferBinding[i].SizeInBytes = output.byteLength;
+            outputBindingDesc[i] = {DML_BINDING_TYPE_BUFFER, &outputBufferBinding[i]};
+            outputOffset = outputOffset + output.byteLength;
+        }
+        mBindingTable->BindOutputs(outputBindingDesc.size(), outputBindingDesc.data());
+
+        // Record execution of the compiled operator.
+        ID3D12DescriptorHeap* descriptorHeaps[] = {mDescriptorHeap.Get()};
+        mCommandList->SetDescriptorHeaps(ARRAYSIZE(descriptorHeaps), descriptorHeaps);
+        mCommandRecorder->RecordDispatch(mCommandList.Get(), compiledGraph->compiledOperator.Get(),
+                                         mBindingTable.Get());
+
+        // Copy buffer from outputResource to readBackResource.
+        CopyBufferRegion(mOutputResource, mReadBackResource, mOutputResourceSize,
+                         D3D12_RESOURCE_STATE_COPY_SOURCE, false);
+        CloseExecuteResetWait();
+
+        D3D12_RANGE tensorBufferRange{0, mOutputResourceSize};
+        int8_t* readBackBuffer;
+        WEBNN_CHECK(mReadBackResource->Map(0, &tensorBufferRange,
+                                           reinterpret_cast<void**>(&readBackBuffer)));
+
+        uint64_t offset = 0;
+        for (size_t i = 0; i < outputNodes.size(); ++i) {
+            offset = RoundUpToMultiple(offset, (uint64_t)DML_MINIMUM_BUFFER_TENSOR_ALIGNMENT);
+            ArrayBufferView output = outputArrayBufferViews[i];
+            if (output.buffer) {
+                memcpy(static_cast<int8_t*>(output.buffer) + output.byteOffset,
+                       readBackBuffer + offset, output.byteLength);
+            }
+            offset += output.byteLength;
+        }
+
+        mReadBackResource->Unmap(0, nullptr);
```

Comment:
So can I remove this unmap() as your suggestion for uploadResource? @bbernhar

Reply:
That's correct. But keep in mind, Map() and Unmap() must be called the same number of times (it is ref-counted by D3D12).